src/ai/NetworkInterface.js

```diff
@@ -0,0 +1,136 @@
+import { credentials } from './APIKeys.js';
+
+/**
+ * @class NetworkInterface
+ * Provides a central location to manage the various fetches made to the toolbox edge server to interact with the
+ * /ai routes on the server, which in turn interact with Azure/OpenAI APIs.
+ * 
+ * @todo: the fetch paths have been temporarily adjusted to work with localhost and locally provided keys,
+ *   and the requests may need to be changed again if you are working with the cloud proxy
```

Comment:
Is there a simple way to swap to the cloud proxy approach?

Reply:
@Steve-KX-RL I forget, what was the difference with this version that you sent me vs. the cloud proxy approach? I remember you sent me a branch that differed from main, and made the requests work on localhost, but I'm not sure what the difference was or how to go back to the cloud proxy version anymore

Reply:
There's 2 major differences between cloud proxy version (the current main edge-server & user-interface branch) and your version:
1. In index.js / wherever starts up & initialize the api keys, cloud-proxy version doesn't send the api keys from user-interface, aka the '/ai/init' route is not called from the user-interface.
2. In cloud-proxy version, Dan and I had a special route to post questions to AI:
            realityEditor.network.postQuestionToAI = async function(conversation, extra) {
                let route = '/ai/question';
                
                const response = await fetch(window.location.origin + route, {
                    headers: {
                        'Accept': 'application/json',
                        'Content-Type': 'application/json'
                    },
                    method: 'POST',
                    body: JSON.stringify({
                        conversation: conversation,
                        extra: extra,
                    })
                });
                const res = await response.json();
                if (res.tools !== undefined) {
                    realityEditor.ai.getToolAnswer(res.category, res.tools);
                } else {
                    realityEditor.ai.getAnswer(res.category, res.answer);
                }
            }
Notice "window.location.origin + route", it's different from the user-interface version.
There might be other small hiccups but these are the main differences.

Reply:
Dan also provided the methods if we want to develop AI feature locally, if we're in cloud-proxy version:
- [ ] When developing locally for the ai chatbot, or any new network requests for the system now:
- [ ] We need edge server, cloud proxy, edge agent add-on, and language server branches all working
- [ ] Run edge server: nvm use 18.17.1 —> npm start
- [ ] Run cloud proxy: nvm use 18.17.1 —> node index.js
- [ ] Run language service: nvm use 18.17.1 —> npm run start
- [ ] Start up edge server —> manage hardware interface —> edgeAgent —>
- [ ] “Enter your cloud proxy host” —> enter “localhost:8000” —> hit “connect”
- [ ] Copy the link in web-desktop client URL, and paste in the browser address tab, make sure to delete “https://“ here, b/c browser might automatically add it

- [ ] Edge agent add-on basically re-directs my request to “[alpha.spatial.ptc.io](http://alpha.spatial.ptc.io/)” / anything that I put in the edge agent “enter your cloud proxy host” box
- [ ] When I put “localhost:8000” in “enter your cloud proxy host”, it means that it reroutes all the network requests on the cloud back to my localhost:8000, so that I can develop new features locally
- [ ] But if I’m not developing new network capabilities (specifically not any new ai chatbot capabilities), I can shut down the cloud proxy, language service, and in the manage-hardware interface —> cloud proxy host —> I should input “[alpha.spatial.ptc.io](http://alpha.spatial.ptc.io/)” (instead of “[spatial.ptc.io](http://spatial.ptc.io/)” so that the cloud branch is somewhat up-to-date with my local branch.
- [ ] Note that default is [spatial.ptc.io](http://spatial.ptc.io/)

Reply:
I'm not sure if this covers everything. @dangond-ptc could you double check if I miss anything?

Reply:
The specific versions of node aren't needed, just as long as it's recent enough, but looks right to me. The Edge Agent allows you to connect via the cloud proxy, which proxies requests through your local edge server but also handles AI requests by sending those to the language server.